lib/bloc/podcast_history/podcast_history_bloc.dart

```diff
@@ -0,0 +1,202 @@
+
+  Stream<PodcastHistorySummationValues> get podcastHistorySumValues =>
+      _podcastHistorySumValues.stream;
+
+  final BehaviorSubject<List<PodcastHistoryModel>> _podcastHistoryList =
+      BehaviorSubject();
+
+  Stream<List<PodcastHistoryModel>> get podcastHistoryList =>
+      _podcastHistoryList.stream;
+
+  final BehaviorSubject<PodcastHistorySortOptions> _podcastHistorySortOption =
+      BehaviorSubject();
+
+  Stream<PodcastHistorySortOptions> get podcastHistorySortOption =>
+      _podcastHistorySortOption.stream;
+
+  //A buffer list comes in use while sorting
+  List<PodcastHistoryModel> bufferPodcastHistoryList = [];
+
+  PodcastHistoryBloc() {
+    registerAsyncHandlers({
+      UpdatePodcastHistoryTimeRange: _updatePodcastHistoryTimeRange,
+      UpdatePodcastHistorySort: _updateSortOption
+    });
+    listenActions();
+
+    _loadSelectedReportTimeRange();
+    _initializeSortOption();
+  }
+
+  void _loadSelectedReportTimeRange() async {
+    PodcastHistoryTimeRange timeRange =
+        await getPodcastHistoryTimeRageFromPrefs();
+
+    _podcastHistoryRange.add(timeRange);
+  }
+
+  void _initializeSortOption() async {
+    PodcastHistorySortOptions sortOption =
+        await getPodcastHistorySortOptionFromPrefs();
+
+    _podcastHistorySortOption.add(sortOption);
+  }
+
+  void _updateSortOption(UpdatePodcastHistorySort action) async {
+    final prefs = await ServiceInjector().sharedPreferences;
+    prefs.setString(_podcastHistorySortKey, action.sortOptions.toJson());
+
+    _podcastHistorySortOption.add(action.sortOptions);
+  }
+
+  Future<PodcastHistorySortOptions>
+      getPodcastHistorySortOptionFromPrefs() async {
+    final prefs = await ServiceInjector().sharedPreferences;
+    PodcastHistorySortOptions sortOption;
+    if (prefs.containsKey(_podcastHistorySortKey)) {
+      sortOption = PodcastHistorySortOptions.fromJson(
+        prefs.getString(_podcastHistorySortKey),
+      );
+    } else {
+      sortOption = PodcastHistorySortOptions.recentlyHeard();
+    }
+    return sortOption;
+  }
+
+  Future<PodcastHistoryTimeRange> getPodcastHistoryTimeRageFromPrefs() async {
+    final prefs = await ServiceInjector().sharedPreferences;
+    PodcastHistoryTimeRange timeRange;
+    if (prefs.containsKey(_podcastHistoryTimeRangeKey)) {
+      timeRange = PodcastHistoryTimeRange.fromJson(
+        prefs.getString(_podcastHistoryTimeRangeKey),
+      );
+    } else {
+      timeRange = PodcastHistoryTimeRange.monthly();
+    }
+    return timeRange;
+  }
+
+  Future _updatePodcastHistoryTimeRange(
+    UpdatePodcastHistoryTimeRange action,
+  ) async {
+    final prefs = await ServiceInjector().sharedPreferences;
+    prefs.setString(_podcastHistoryTimeRangeKey, action.range.toJson());
+    await fetchPodcastHistory(
+        startDate: action.range.startDate, endDate: action.range.endDate);
+
+    _podcastHistoryRange.add(action.range);
+  }
+
+  Future<List<PodcastHistoryModel>> fetchPodcastHistory(
```

Comment:
I think it is better to use the sqlite engine to return the result set grouped by podcast id with the totals related to the podcast. the number of payments is far greater than the number of podcasts listened to so it would minimize a lot the rows returned.
We can also do the same for sorting but it is a nice to have, not critical.

Reply:
@roeierez Sure. For sorting, we are not reading from local DB we are simply modifying the already fetched list.

Reply:
Done